### PR TITLE
User project comments

### PR DESCRIPTION
**Pull request by andreamlin**
Addresses issue #2494.
Add comments to methods that accept `userProject` parameter.

~Blocked on #2558 .~

### CODE REVIEW

**Comment by andreamlin** [comment]
@frankyn There's method comments in the changelist for Storage.java. Is there anything else you were looking for?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 1]
```diff
@@ -834,6 +834,15 @@ public static BlobListOption currentDirectory() {
     }
 
     /**
+     * Returns an option for bucket's billing user project. This option is only used by the buckets with
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: This option is required by buckets with `requester_pays` flag enabled.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 2]
```diff
@@ -1378,6 +1387,8 @@ public static Builder newBuilder() {
   /**
    * Creates a new bucket.
    *
+   * Accepts an optional userProject {@link BucketTargetOption} option.
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Accepts an optional userProject {@link BucketTargetOption} option which defines the project id to assign operational costs.

**Comment by frankyn** [comment]
Thanks @andreamlin! I added comments to clarify that UserProject can still be used without Requester Pays enabled for a bucket. A user can use it to assign operational costs to a specific project as long as the service account has roles/Owner or roles/Editor.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 1 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Done.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 2 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Done.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java [hunk 3]
```diff
@@ -697,11 +697,6 @@ public BucketAccessControl getAcl(String bucket, String entity, Map<Option, ?> o
   }
 
   @Override
-  public BucketAccessControl getAcl(String bucket, String entity) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: Shouldn't this method still exist and be provided as an overloaded method option?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 4]
```diff
@@ -1387,7 +1387,8 @@ public static Builder newBuilder() {
   /**
    * Creates a new bucket.
    *
-   * Accepts an optional userProject {@link BucketTargetOption} option.
+   * Accepts an optional userProject {@link BucketTargetOption} option which defines the project id
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Could you apply this update to all instances of this comment?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 5]
```diff
@@ -834,8 +834,8 @@ public static BlobListOption currentDirectory() {
     }
 
     /**
-     * Returns an option for bucket's billing user project. This option is only used by the buckets with
-     * 'requester_pays' flag.
+     * Returns an option for bucket's billing user project. This option is required for and only
+     * used by the buckets with 'requester_pays' flag.
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Can you update this to:
```
Returns an option to define the billing user project. This option is required by buckets with `requester_pays` flag enabled to assign operation costs.
```

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java (hunk 3 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: In the underlying diffbase PR #2558 , we decided to just add a parameter to the method signature instead of overloading it, even though this is a breaking change.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 4 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Right right :)

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java (hunk 3 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: I'd push not to cause breaking changes to the GA library and instead overloading.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 5 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Done.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/StorageRpc.java [hunk 6]
```diff
@@ -413,19 +415,19 @@ void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset,
...
-  Policy setIamPolicy(String bucket, Policy policy);
+  Policy setIamPolicy(String bucket, Policy policy, Map<Option, ?> options);
 
   /**
    * Tests whether the caller holds the specified permissions for the specified bucket.
    *
    * @throws StorageException upon failure
    */
-  TestIamPermissionsResponse testIamPermissions(String bucket, List<String> permissions);
+  TestIamPermissionsResponse testIamPermissions(String bucket, List<String> permissions, Map<Option, ?> options);
```
[the remark this answers is not in the document]
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/StorageRpc.java]: not a test.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java [hunk 7]
```diff
@@ -69,6 +69,7 @@
 import java.lang.reflect.Field;
 import java.math.BigInteger;
 import java.util.ArrayList;
+import java.util.HashMap;
```
[the remark this answers is not in the document]
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: Done.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java (hunk 3 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: https://github.com/GoogleCloudPlatform/google-cloud-java/pull/2558#discussion_r148646550
tl;dr overloading the method is a breaking change anyhow.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/StorageRpc.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/StorageRpc.java]: NOT A TEST

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 8]
```diff
@@ -834,6 +834,15 @@ public static BlobListOption currentDirectory() {
     }
 
     /**
+     * Returns an option for bucket's billing user project. This option is required for and only
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: This comment doesn't match comment above.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java (hunk 3 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: IIUC, this only breaks code that overloaded the interface, but I'm not sure how often users overload the interface themselves versus not. I'd still vote to add overloaded methods.

Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: The StorageRpc (and this HttpStorageRpc) isn't really intended to be used by the general user. We'd assume they would use the Storage class, which wraps the StorageRpc calls in retry and exception-handling logic.

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: Got it, thanks!

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 8 above)
Comment (andreamlin) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Done.

**Comment by frankyn** [comment]
Thank you! @andreamlin

**Comment by andreamlin** [comment]
@neozwu PTAL
